src/events/google_cloud.js

```diff
@@ -79,6 +86,142 @@ function bigQueryWrapper(wrappedFunction) {
+                    }
+                };
+            });
+            tracer.addEvent(pubsubEvent, responsePromise);
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [config, patchedCallback]);
+    };
+}
+
+/**
+ * Wrap pubsub request function.
+ * @param {Function} original pubsub request function.
+ * @returns {Function} The wrapped function
+ */
+function wrapPubSubPullFunction(original) {
+    return function internalPubSubPullFunction(request, options, callback) {
+        let clientPromiseRequest;
+        try {
+            let pubsubProjectId;
+            if (request && request.subscription) {
+                const subscriptionSplited = request.subscription.split('/');
+                if (subscriptionSplited.length > 1) {
+                    [, pubsubProjectId] = subscriptionSplited;
+                }
+            }
+            const { slsEvent: pubsubEvent, startTime } = eventInterface.initializeEvent(
+                GOOGLE_CLOUD_TYPES.pubsub.name, pubsubProjectId, 'Pull', GOOGLE_CLOUD_TYPES.pubsub.name
+            );
+            const patchedCallback = (err, res) => {
+                const responseMetadata = {};
+                if (res && res.receivedMessages) {
+                    const receivedMessages = res.receivedMessages.reduce((acc, current) => {
+                        acc.push({ messageId: current.message.messageId, message: `${current.message.data}` });
+                        return acc;
+                    }, []);
+                    responseMetadata.receivedMessages = receivedMessages;
+                }
+                eventInterface.finalizeEvent(pubsubEvent, startTime, err, responseMetadata);
+                if (callback) {
+                    callback(err, res);
+                }
+            };
+            // in case callback has given from the client.
+            if (callback) {
+                let patchedCallbackWithPromise = callback;
+                const promise = new Promise((resolve) => {
+                    patchedCallbackWithPromise = (err, res) => {
+                        resolve();
+                        patchedCallback(err, res);
+                    };
+                });
+                tracer.addEvent(pubsubEvent, promise);
+                return original.apply(this, [request, options, patchedCallbackWithPromise]);
+            }
+            clientPromiseRequest = original.apply(this, [request, options, callback]);
+            tracer.addEvent(pubsubEvent, clientPromiseRequest);
+            clientPromiseRequest.then((res) => {
+                const [response] = res;
+                patchedCallback(null, response);
+                return res;
+            }, (err) => {
+                patchedCallback(err, null);
+                throw err;
+            });
+        } catch (err) {
+            tracer.addException(err);
+        }
+        if (!clientPromiseRequest) {
```

Comment:
I think you should put this code in the catch (that's the only way it can run), will be more readable

Reply:
If tracer.addEvent throwing an error, and clientPromiseRequest was called - I don't want to call the original function twice.